fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/simplification/Simplification.java

```diff
@@ -0,0 +1,176 @@
+
+        final var newRoot = root.<Value>mapMaybe((current, mappedChildren) -> {
+            current = computeCurrent(current, mappedChildren);
+            return executeRuleSet(root,
+                    current,
+                    ruleSet,
+                    (rule, r, c, plannerBindings) -> new ValueComputationRuleCall<>(rule, r, c, argument, plannerBindings, aliasMap, constantAliases, resultsMap::get),
+                    results -> onResultsFunction(resultsMap, results));
+        }).orElseThrow(() -> new RecordCoreException("expected a mapped tree"));
+        return resultsMap.get(newRoot);
+    }
+
+    @Nonnull
+    @SuppressWarnings("PMD.CompareObjectsWithEquals")
+    private static Value computeCurrent(@Nonnull final Value current, @Nonnull final Iterable<? extends Value> mappedChildren) {
+        final var children = current.getChildren();
+        final var childrenIterator = children.iterator();
+        final var mappedChildrenIterator = mappedChildren.iterator();
+        boolean isSame = true;
+        while (childrenIterator.hasNext() && mappedChildrenIterator.hasNext()) {
+            final Value child = childrenIterator.next();
+            final Value mappedChild = mappedChildrenIterator.next();
+            if (child != mappedChild) {
+                isSame = false;
+                break;
+            }
+        }
+        // make sure they are both exhausted or both are not exhausted
+        Verify.verify(childrenIterator.hasNext() == mappedChildrenIterator.hasNext());
+        return isSame ? current : current.withChildren(mappedChildren);
+    }
+
+    @Nonnull
+    private static <R> Value onResultsFunction(@Nonnull final Map<Value, ValueComputationRuleCall.ValueWithResult<R>> resultsMap,
+                                               @Nonnull final Collection<ValueComputationRuleCall.ValueWithResult<R>> results) {
+        Verify.verify(results.size() <= 1);
+
+        final var valueWithResult = Iterables.getOnlyElement(results);
+        final var value = valueWithResult.getValue();
+        resultsMap.put(value, valueWithResult);
+        return value;
+    }
+
+    @Nonnull
+    @SuppressWarnings("PMD.CompareObjectsWithEquals")
+    private static <R, C extends AbstractValueRuleCall<R, C>> Value executeRuleSet(@Nonnull final Value root,
+                                                                                   @Nonnull Value current,
+                                                                                   @Nonnull final AbstractValueRuleSet<R, C> ruleSet,
+                                                                                   @Nonnull final RuleCallCreator<R, C> ruleCallCreator,
+                                                                                   @Nonnull final Function<Collection<R>, Value> onResultsFunction) {
+        final boolean isRoot = current == root;
+        Value newCurrent = current;
+        do {
```

Comment:
does it make sense to add an upper bound of iterations to safe-guard against (in)directly-cyclic rules, i.e.:

rule1: v1 -> v2
rule2: v2 -> v3
rule3: v3 -> v4
...
rulen: vn -> v1

Reply:
Hmm, we don't have a good way of configuring that although it is useful. On the other hand, you can make the case that simplification is monotonic (or should be). I think we can agree that this should only be possible if there is an actual bug in the rules, not if the candid `Value` is too complex (which can be the reason for the Cascades plane). Let me think about how to do that.

Reply:
Maybe a relatively high total time limit on planning? Just to avoid stuck threads due to bugs.

Reply:
Opened https://github.com/FoundationDB/fdb-record-layer/issues/1854 to track this general problem.